[Feature/Entity-DTO 구현] 처리서버 Entity, DTO,  Controller, Service 구현 및 테스트 클래스 작성

ongsttt52 opened the pull request:
### Summary

### Description

<!-- 구체적인 작업 내용, 필요시 이미지 첨부 -->
- 주기정보, 시동정보 Entity 및 DTO 구현

- CollectController, CollectService 클래스 작성

- CycleInfo, EventInfo 및 CycleData를 저장하는 로직 구현

- 중복 데이터가 저장될 경우 200 OK를 반환하도록 예외처리

- 위 서비스 로직의 테스트 케이스 작성

### Discussion

<!-- 추후 논의할 점 -->
- Service 메서드 테스트 방식 (Mocking, 인메모리 DB, Transactional)

- 공통 응답 객체 및 GlobalExceptionHandler 구조

### Issue Number or Link

<!-- 예시: closes #issue-number -->
closes #25 #12

ongsttt52 committed:
chore: build.gradle 수정

루트 모듈, 서브 모듈 build.gradle에 bootJar, jar { enabled } 옵션 설정

ongsttt52 committed:
feat: collect API 구현

-[] CycleInfo, CycleData Entity, DTO 구현
-[] 이벤트 정보, 주기 정보를 DB에 저장하는 로직 구현

ongsttt52 committed:
feat: 주기정보 테이블 제약조건 추가

-[] cycle_info 테이블 mdn(차량번호)과 o_time, on_time, off_time 컬럼을 각각 유니크 제약조건으로 설정

ongsttt52 committed:
feat: 주기정보, 이벤트정보 중복 저장 예외처리 추가

-[] 주기정보가 데이터베이스에 저장은 되었지만 응답 전송이 실패한 경우를 위한 로직 추가
-[] 주기정보, 이벤트정보가 저장될 때 차량번호(mdn)와 시간이 중복되면 예외를 발생시키고 HttpCode 200을 반환하도록 함

ongsttt52 committed:
chore: 빌드 스크립트 중복된 코드 제거

-[] build.gradle의 bootJar, jar가 중복 작성되어 있어 제거함

ongsttt52 committed:
feat: Controller, Service 구현 및 테스트 클래스 작성

-[] CollectController, CollectService 클래스 작성
-[] CycleInfo, EventInfo 및 CycleData를 저장하는 로직 구현
-[] 중복 데이터가 저장될 경우 200 OK를 반환하도록 예외처리
-[] 위 서비스 로직의 테스트 케이스 작성

yyysolhhh commented on build.gradle:
```diff
@@ -60,4 +60,4 @@ bootJar {
 jar {
     enabled = false
-}
+}
```
빨간 표시돼있는 부분은 마지막 빈줄을 안넣어서 생기는거라 이런 부분들 모두 마지막에 newline 넣어주세요.

kimchijinju commented on charlee-collector/src/main/java/com/efa73/charleecollector/interfaces/controller/CollectController.java:
```diff
@@ -0,0 +1,53 @@
+
+        String mdn = collectService.collectCycle(cycleInfo, cycleDataList);
+
+        return CollectResponse.builder()
```
왜 Record가 아니라 Class를 사용하셨나요?
이 상황에서는 Record로 만들어도 될듯 합니다

yyysolhhh commented on charlee-collector/src/main/java/com/efa73/charleecollector/domain/entity/CycleInfo.java:
```diff
@@ -0,0 +1,60 @@
+                @UniqueConstraint(name = "uq_cycleinfo_mdn_offtime", columnNames = {"mdn","off_time"})
+        }
+)
+@Getter @NoArgsConstructor @AllArgsConstructor @Builder
```
여러줄로 나눠서 하면 보기 좋을 것 같아요.

kimchijinju commented on charlee-collector/src/main/java/com/efa73/charleecollector/domain/entity/CycleData.java:
```diff
@@ -0,0 +1,59 @@
+@Getter
+@NoArgsConstructor
+@AllArgsConstructor
+@Builder
```
Entity에 Builder가 꼭 필요할까요?
Entity는 무결한 객체로 만들어 진 다음 DB에 저장 되어야 데이터 관리가 쉬워질 것 같습니다.
빌더를 사용하면 완전한 객체가 만들어지지 않을 가능성이 큽니다.
팩터리 메서드나 생성자를 통해 받는 데이터를 강제하는 것이 어떨까요?

yyysolhhh commented on charlee-collector/src/main/java/com/efa73/charleecollector/domain/service/CollectService.java:
```diff
@@ -0,0 +1,43 @@
+
+            return saved.getMdn();
+        } catch (DataIntegrityViolationException e) {
+            return "중복저장됨";
```
ExceptionHandler를 만드는건 어떤가요?

kimchijinju replied:
이 부분에 대해서 처리가 애매한 것 같습니다.

Todo로 두고, 추후에 부하테스트 해보며 처리하도록 합시다.

kimchijinju commented on charlee-collector/src/main/java/com/efa73/charleecollector/domain/entity/CycleInfo.java:
```diff
@@ -0,0 +1,60 @@
+@Getter @NoArgsConstructor @AllArgsConstructor @Builder
+public class CycleInfo {
+
+    public static final int DATA_LIST_LENGTH = 60;
```
서버 로직에서는 해당 데이터는 없어도 무방해 보입니다.

yyysolhhh commented on charlee-collector/src/main/java/com/efa73/charleecollector/interfaces/controller/CollectController.java:
```diff
@@ -0,0 +1,53 @@
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+@RequestMapping("/collector")
```
/api 접두사 붙여주세요.

yyysolhhh commented on charlee-collector/src/test/java/com/efa73/charleecollector/domain/entity/CollectTestFactory.java:
```diff
@@ -0,0 +1,57 @@
+import org.springframework.stereotype.Component;
+
+@Component
+@ComponentScan
```
ComponentScan은 다른 패키지에 있는 빈을 스캔할때 쓰는거라고 알고있는데 여기서 꼭 필요한 건가요?

kimchijinju commented on charlee-collector/src/main/java/com/efa73/charleecollector/interfaces/dto/request/CycleDataRequest.java:
```diff
@@ -0,0 +1,38 @@
+) {
+
+    public static CycleData toEntity(CycleDataRequest cycleDataRequest, CycleInfo cycleInfo) {
+        return CycleData.builder()
```
마찬가지로 Builder 패턴을 지양하면 좋을 듯 합니다.

yyysolhhh commented on charlee-collector/src/test/java/com/efa73/charleecollector/domain/entity/CollectTestFactory.java:
```diff
@@ -0,0 +1,57 @@
+
+@Component
+@ComponentScan
+public class CollectTestFactory {
```
이 클래스는 테스트코드가 아닌데 test 하위에 있어요. 위치가 잘못된것 같습니다.

yyysolhhh commented on charlee-collector/src/test/java/com/efa73/charleecollector/domain/service/CollectServiceTest.java:
```diff
@@ -0,0 +1,102 @@
+    }
+
+    @Test
+    // collectCycle()에서 CycleInfo가 잘 저장되고 mdn을 반환함을 검증
```
이 부분은 DisplayName으로 제목을 써주면 좋을것 같아요.

yyysolhhh commented on charlee-collector/src/test/java/com/efa73/charleecollector/domain/service/CollectServiceTest.java:
```diff
@@ -0,0 +1,102 @@
+    }
+
+    @Test
+    // collectCycle()에서 cycle_info 테이블에 중복 저장이 실행될 경우 예외처리 로직을 검증
```
이부분도 마찬가지로 DisplayName을 써주세요.

yyysolhhh commented on charlee-collector/src/test/resources/application.yml:
```diff
@@ -0,0 +1,37 @@
+
+springdoc:
+  swagger-ui:
+    path: /swagger-ui
```
test에 사용할 application.yml에는 불필요한 설정이 많은 것 같아요. 꼭 필요한 설정만 넣어주세요.

yyysolhhh reviewed (commented):
1. 파일 마지막에 빈줄 다 넣어주세요.
2. 테스트코드에서 주석으로 설명 다는 것보다 DisplayName에 한글로 쓰는게 좋을것 같아요.
3. 그리고 다시 올리실때 ./gradlew build -x test 통과되는지 확인해주세요.

kimchijinju reviewed (changes requested):
확인했습니다.
보고 수정하시면 다시 리뷰남길게요

ongsttt52 committed:
refactor: builder 패턴 교체 및 테스트 클래스 수정

-[] Entity, DTO의 builder 전부 제거, 정적 팩토리 메서드로 교체
-[] CollectTestFactory 클래스 삭제, @BeforeEach 내부에서 데이터 생성하는 방식으로 변경
-[] test에서 사용중인 application.yml의 불필요한 설정 제거
-[] 기타 불필요한 어노테이션, 주석 등 제거

ongsttt52 commented:
> 1. 파일 마지막에 빈줄 다 넣어주세요.
> 2. 테스트코드에서 주석으로 설명 다는 것보다 DisplayName에 한글로 쓰는게 좋을것 같아요.
> 3. 그리고 다시 올리실때 ./gradlew build -x test 통과되는지 확인해주세요.

1,2 번 사항 수정하였고 빌드 테스트도 성공 확인 후 다시 올렸습니다.

kimchijinju commented:
담부터는 커밋 단위만 좀더 쪼개주시면 될 것 같습니다!
